wgengine/netstack/netstack.go

```diff
@@ -37,161 +39,241 @@ import (
+		linkEP:  linkEP,
+		tundev:  tundev,
+		e:       e,
+		mc:      mc,
 	}
+	return ns, nil
+}
 
-	e.AddNetworkMapCallback(func(nm *netmap.NetworkMap) {
-		oldIPs := make(map[tcpip.Address]bool)
-		for _, ip := range ipstack.AllAddresses()[nicID] {
-			oldIPs[ip.AddressWithPrefix.Address] = true
+// Run sets up all the handlers so netstack can start working. Implements
+// wgengine.FakeImpl.
+func (ns *Impl) Run() error {
+	ns.e.AddNetworkMapCallback(ns.updateIPs)
+	tcpFwd := tcp.NewForwarder(ns.ipstack, 0, 16, ns.acceptTCP)
+	udpFwd := udp.NewForwarder(ns.ipstack, ns.acceptUDP)
+	ns.ipstack.SetTransportProtocolHandler(tcp.ProtocolNumber, tcpFwd.HandlePacket)
+	ns.ipstack.SetTransportProtocolHandler(udp.ProtocolNumber, udpFwd.HandlePacket)
+	go ns.injectOutbound()
+	ns.tundev.PostFilterIn = ns.injectInbound
+
+	return nil
+}
+
+func (ns *Impl) updateIPs(nm *netmap.NetworkMap) {
+	oldIPs := make(map[tcpip.Address]bool)
+	for _, ip := range ns.ipstack.AllAddresses()[nicID] {
+		oldIPs[ip.AddressWithPrefix.Address] = true
+	}
+	newIPs := make(map[tcpip.Address]bool)
+	for _, ip := range nm.Addresses {
+		newIPs[tcpip.Address(ip.IP.IPAddr().IP)] = true
+	}
+
+	ipsToBeAdded := make(map[tcpip.Address]bool)
+	for ip := range newIPs {
+		if !oldIPs[ip] {
+			ipsToBeAdded[ip] = true
 		}
-		newIPs := make(map[tcpip.Address]bool)
-		for _, ip := range nm.Addresses {
-			newIPs[tcpip.Address(ip.IPNet().IP)] = true
+	}
+	ipsToBeRemoved := make(map[tcpip.Address]bool)
+	for ip := range oldIPs {
+		if !newIPs[ip] {
+			ipsToBeRemoved[ip] = true
 		}
+	}
 
-		ipsToBeAdded := make(map[tcpip.Address]bool)
-		for ip := range newIPs {
-			if !oldIPs[ip] {
-				ipsToBeAdded[ip] = true
-			}
+	for ip := range ipsToBeRemoved {
+		err := ns.ipstack.RemoveAddress(nicID, ip)
+		if err != nil {
+			ns.logf("netstack: could not deregister IP %s: %v", ip, err)
+		} else {
+			ns.logf("[v2] netstack: deregistered IP %s", ip)
 		}
-		ipsToBeRemoved := make(map[tcpip.Address]bool)
-		for ip := range oldIPs {
-			if !newIPs[ip] {
-				ipsToBeRemoved[ip] = true
-			}
+	}
+	for ip := range ipsToBeAdded {
+		var err *tcpip.Error
+		if ip.To4() == "" {
+			err = ns.ipstack.AddAddress(nicID, ipv6.ProtocolNumber, ip)
+		} else {
+			err = ns.ipstack.AddAddress(nicID, ipv4.ProtocolNumber, ip)
 		}
-
-		for ip := range ipsToBeRemoved {
-			err := ipstack.RemoveAddress(nicID, ip)
-			if err != nil {
-				logf("netstack: could not deregister IP %s: %v", ip, err)
-			} else {
-				logf("netstack: deregistered IP %s", ip)
-			}
+		if err != nil {
+			ns.logf("netstack: could not register IP %s: %v", ip, err)
+		} else {
+			ns.logf("[v2] netstack: registered IP %s", ip)
 		}
-		for ip := range ipsToBeAdded {
-			err := ipstack.AddAddress(nicID, ipv4.ProtocolNumber, ip)
-			if err != nil {
-				logf("netstack: could not register IP %s: %v", ip, err)
-			} else {
-				logf("netstack: registered IP %s", ip)
-			}
+	}
+}
+
+func (ns *Impl) dialTCP(address string) (*gonet.TCPConn, error) {
```

Comment:
no context.Context here? how long can this block? how do we cancel this?

Reply:
Added context, let me know if I'm missing any other place where a context would be required in this PR please.